src/Tempest/Http/src/Routing/Construction/RoutingTree.php

```diff
@@ -0,0 +1,36 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Tempest\Http\Routing\Construction;
+
+/**
+ * @internal
+ */
+final class RoutingTree
```

Comment:
So your routing tree contains one RouteTreeNode per routing method. 

For each `add` call we check if we need to create a new Root Node. And add the route to it by means of recursion.

The tree can also return all routing regexes by means of recursion .